pythoncode/usbTrainer.py

```diff
@@ -92,477 +101,826 @@
+    # C a l i b r a t e S u p p o r t e d
+    #---------------------------------------------------------------------------
+    # input     self.tt
+    #
+    # function  Return whether this trainer support calibration
+    #
+    # returns   True/False
+    #---------------------------------------------------------------------------
+    def CalibrateSupported(self):
+        if self.Headunit == hu1932:                 # And perhaps others as well
+            return True
+        else:
+            return False
 
-    rtn = ResistanceAtLowSpeed(SpeedKmh, rtn)
+    #---------------------------------------------------------------------------
+    # Convert Grade (slope) to power
+    #
+    # TotalReverse:
+    #   documents:      LOAD ~= (slope (in %) - 0.4) * 650
+    #   ttyT1941.py:    SlopeScale =  2 * 5 * 130    = 1300
+    #                   commented  = 13 * 5 * 10 * 5 = 3250
+    #
+    # TotalReverse, Issue #2:
+    # ANT+ describes a simple physical model:
+    #   Total resistance [N] = Gravitational Resistance + Rolling Resistance + Wind Resistance
+    # Gravitational Resistance [N] = (Equipment Mass + User Mass) x Grade(%)/100 x 9.81
+    # With the assumption of a total weight of about 80 kg and without rolling and wind
+    # resistance together with the simple assumption
+    # Load = Force [N] * 137 (the result of a fitting in ergo mode) one get:
+    #
+    # Load = 137 * 80 * 9.81 * slope_in_percent / 100
+    #
+    # SlopeScale = 137 * 80 * 9.81 / 100 = 1075
+    #
+    # Tests:
+    # factor 650 causes Zwift to think we ride @ 20km/h where the wheel runs at 40km/hr
+    # factor 1300 to be tested
+    #
+    #---------------------------------------------------------------------------
+    # Refer to ANT msgUnpage51_TrackResistance()
+    #          where zwift grade does not seem to match the definition
+    #---------------------------------------------------------------------------
+    # Input         TargetGrade, UserAndBikeWeight, SpeedKmh
+    #
+    # Function      Calculate what power must be produced, given the current
+    #               grade, speed and weigth
+    # 
+    # Output        TargetPower
+    #---------------------------------------------------------------------------
+    def Grade2Power(self):                                         #Grade2Power#
+        if self.TargetGrade < -20 or self.TargetGrade > 30:
+            logfile.Console("Grade2Power; TargetGrade (%s) out of bounds!" % (self.TargetGrade))
+
+        if self.UserAndBikeWeight < 70:
+            self.UserAndBikeWeight = 75 + 10
 
-    return rtn
+        self.TargetPower  = self.__PfietsNL()
 
-def PfietsNL(TargetGrade, UserAndBikeWeight, SpeedKmh):
+        if debug.on(debug.Application):
+            logfile.Write ("Grade2Power; TargetGrade=%4.1f%%, Speed=%4.1f, Weight=%3.0f, TargetPower=%3.0fW" % \
+                (self.TargetGrade, self.SpeedKmh, self.UserAndBikeWeight, self.TargetPower) )
+
+    def __PfietsNL(self):

```

Comment:
I would argue this should not be a member function of a trainer class. It is a function that implements a physical calculation, nothing specific to a trainer - not a function on a trainer.

It is parameterised by several things - current speed (a property of a trainer), user and bike weight (not a property of a trainer), target grade (not a property of a trainer), and all the various ANT page tuning parameters (of which weight and grade are two, drag coefficients etc the others). I get that having it as a member function means some of those variables are available on self., but that's convenience at the expense of complexity.

My aim would be to simplify the trainer class hierarchy to a bare minimum - to only have in there what is needed for a common interface and what is needed to implement trainer-specifics (message types, trainer-specific unit conversions etc).

To me, "code coupling" (modification variables inside functions, distributed logic, etc...) weaves complexity into the code - it takes longer to read and understand, change or re-use. In contrast, decoupling reduces complexity and creates clarity, where updates to state are explicit, functions are pure(r), and nothing important is hidden away. The logic that actually does the interesting work should be separated from the machinery that implements it, and then be plainly stated and as close together as possible.

As an aside: it's why calling SetGrade() from inside the ANT+ handling routine irks me - why I turned it around so that rather than the code saying "the act of receiving an ANT page immediately changes trainer state", this become three much cleaner, decoupled steps, namely "the act of receiving an ANT page updates ANT page state", "TargetGrade is a function of ANT page state" and then "change trainer state".

It is down to style. I'm not saying "do what I did", just describing my thought process. It's a subtle change of approach but has profound impact to maintainability in my experience. My mantra - "decouple, decouple, decouple". My other mantra is "const-correctness" but as we discussed, Python doesn't have const.

Reply:
I agree on approach, but it's different from what I did.

clsTacxTrainer is not only the USB-interface but the entire trainer, where required power or grade can be set (SetPower(), SetGrade()). Other design decisions could have been made, but this is how it's done.
Also clsTacxTrainer does not know about ANT and/or FE-C (sorry for devAntDongle which could be removed) and could be used in TTS if Tacx would like (conceptually, of course they don't).

Trainer-specifics have been put in child-classes and this is a great advantage over the previous implementation; clsTacxLegacyUsbTrainer and clsTacxVortexTrainer have not been tested (although I believe code is 98% ok) and when updated later on it's most likely that the now tested code needs not be touched and hence there is no regression.

There were two big issues in FortiusANT:
- too many locations of data, this is resolved
- Grade2Power() did not work; see all issues now open

I am going to focus on the Grade2Power() and virtual speed now, in the meantime improving code where required. Hope you can join in on that